airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlQueryUtils.java

```diff
@@ -0,0 +1,301 @@
+import io.airbyte.cdk.integrations.source.relationaldb.models.InternalModels.StateType;
+import io.airbyte.cdk.integrations.source.relationaldb.state.StateManager;
+import io.airbyte.commons.json.Jsons;
+import io.airbyte.protocol.models.AirbyteStreamNameNamespacePair;
+import io.airbyte.protocol.models.v0.ConfiguredAirbyteCatalog;
+import io.airbyte.protocol.models.v0.ConfiguredAirbyteStream;
+import java.math.BigDecimal;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Utility class to define constants related to querying mssql
+ */
+public class MssqlQueryUtils {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(MssqlQueryUtils.class);
+  private static final String MAX_OC_VALUE_QUERY =
+      """
+        SELECT MAX(%s) as %s FROM %s;
+      """;
+
+  public record TableSizeInfo(Long tableSize, Long avgRowLength) {}
+
+  private static final String MAX_CURSOR_VALUE_QUERY =
+      """
+        SELECT %s FROM %s WHERE %s = (SELECT MAX(%s) FROM %s);
+      """;
+  public static final String INDEX_QUERY = "EXEC sp_helpindex N'%s'";
+
+  public record Index(
+                      @JsonProperty("index_name") String name,
+                      @JsonProperty("index_description") String description,
+                      @JsonProperty("index_keys") String keys) {}
+
+  public static final String TABLE_ESTIMATE_QUERY =
+      """
+      EXEC sp_spaceused N'"%s"."%s"'
+      """;
+
+  public static final String MAX_OC_COL = "max_oc";
+  public static final String DATA_SIZE_HUMAN_READABLE = "data";
+  public static final String NUM_ROWS = "rows";
+
+  public static void getIndexInfoForStreams(final JdbcDatabase database, final ConfiguredAirbyteCatalog catalog, final String quoteString) {
+    for (final ConfiguredAirbyteStream stream : catalog.getStreams()) {
+      final String streamName = stream.getStream().getName();
+      final String schemaName = stream.getStream().getNamespace();
+      final String fullTableName = getFullyQualifiedTableNameWithQuoting(schemaName, streamName, quoteString);
+      LOGGER.info("Discovering indexes for table {}", fullTableName);
+      try {
+        final String query = INDEX_QUERY.formatted(fullTableName);
+        LOGGER.debug("Index lookup query: {}", query);
+        final List<JsonNode> jsonNodes = database.bufferedResultSetQuery(conn -> conn.prepareStatement(query).executeQuery(),
+            resultSet -> JdbcUtils.getDefaultSourceOperations().rowToJson(resultSet));
+        if (jsonNodes != null) {
+          jsonNodes.stream().map(node -> Jsons.convertValue(node, Index.class))
+              .forEach(i -> LOGGER.info(String.valueOf(i)));
+        }
+      } catch (final Exception ex) {
```

Comment:
Since this is informational log I never want it to stop the sync.
There are some variations that are not known - all sorts of hosted and managed and serverless db services won't let you look under the hood.
In the past we ended up adding this to equivalent pre sync logs - getting the version and engine of mysql for example.
Not sure what you meant with `final`